vms/platformvm/txs/mempool/mempool.go

```diff
@@ -5,42 +5,411 @@ package mempool
-	ErrCantIssueAdvanceTimeTx     = errors.New("can not issue an advance time tx")
-	ErrCantIssueRewardValidatorTx = errors.New("can not issue a reward validator tx")
+	ErrGasCapacityExceeded = errors.New("gas capacity exceeded")
+
+	_ txs.Visitor = (*utxoGetter)(nil)
 )
 
+type Tx struct {
+	*txs.Tx
+	Complexity gas.Dimensions
+	GasPrice   float64
+
+	gasUsed gas.Gas
+}
+
 type Mempool struct {
-	txmempool.Mempool[*txs.Tx]
+	weights     gas.Dimensions
+	gasCapacity gas.Gas
+	avaxAssetID ids.ID
+
+	lock               sync.Mutex
+	maxHeap            heap.Map[ids.ID, Tx]
+	minHeap            heap.Map[ids.ID, Tx]
+	consumedUTXOs      *setmap.SetMap[ids.ID, ids.ID]
+	droppedTxIDs       *lru.SizedCache[ids.ID, error]
+	currentGas         gas.Gas
+	numTxsMetric       prometheus.Gauge
+	gasAvailableMetric prometheus.Gauge
 }
 
-func New(namespace string, registerer prometheus.Registerer) (*Mempool, error) {
-	metrics, err := txmempool.NewMetrics(namespace, registerer)
-	if err != nil {
-		return nil, err
+func New(
+	namespace string,
+	weights gas.Dimensions,
+	gasCapacity gas.Gas,
+	avaxAssetID ids.ID,
+	registerer prometheus.Registerer,
+) (*Mempool, error) {
+	numTxsMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Name:      "count",
+		Help:      "number of transactions in the mempool",
+	})
+
+	gasAvailableMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Name:      "gas_available",
+		Help:      "amount of remaining gas available",
+	})
+
+	if err := errors.Join(
+		registerer.Register(numTxsMetric),
+		registerer.Register(gasAvailableMetric),
+	); err != nil {
+		return nil, fmt.Errorf("failed to register metrics: %w", err)
 	}
-	pool := txmempool.New[*txs.Tx](
-		metrics,
-	)
-	return &Mempool{Mempool: pool}, nil
+
+	return &Mempool{
+		weights:     weights,
+		gasCapacity: gasCapacity,
+		avaxAssetID: avaxAssetID,
+		maxHeap: heap.NewMap[ids.ID, Tx](func(a, b Tx) bool {
+			return a.GasPrice > b.GasPrice
+		}),
+		minHeap: heap.NewMap[ids.ID, Tx](func(a, b Tx) bool {
+			return a.GasPrice < b.GasPrice
+		}),
+		consumedUTXOs: setmap.New[ids.ID, ids.ID](),
+		droppedTxIDs: lru.NewSizedCache[ids.ID, error](64, func(ids.ID, error) int {
+			return 1
+		}),
+		numTxsMetric:       numTxsMetric,
+		gasAvailableMetric: gasAvailableMetric,
+	}, nil
 }
 
 func (m *Mempool) Add(tx *txs.Tx) error {
-	switch tx.Unsigned.(type) {
-	case *txs.AdvanceTimeTx:
-		return ErrCantIssueAdvanceTimeTx
-	case *txs.RewardValidatorTx:
-		return ErrCantIssueRewardValidatorTx
-	default:
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	complexity, gasUsed, err := m.meter(tx.Unsigned)
+	if err != nil {
+		return fmt.Errorf("failed to meter tx: %w", err)
+	}
+
+	consumedAVAX, producedAVAX, err := getUTXOs(m.avaxAssetID, tx.Unsigned)
+	if err != nil {
+		return fmt.Errorf("failed to get utxos: %w", err)
+	}
+
+	feesPaid := consumedAVAX - producedAVAX
+	gasPrice := float64(0)
+	if gasUsed == 0 {
+		gasPrice = 0
+	} else {
```

Comment:
This should never happen right? I assume this is because tests don't initialize the weights correctly? Feels like it might make more sense to initialize the weights correctly in all of our tests rather than adding this code here.

Reply:
Yeah this is purely for tests... we can change this to require a non-zero gas used